### DISCUSSION TITLE
Ports and alignment

### DISCUSSION THREAD

**Discussion by sonicsmooth** [terrastruct/d2#2013] [Feature Requests]
I just discovered D2 and love it already.

Adding context/feature requests to:
[https://github.com/terrastruct/d2/discussions/605](url)
[https://github.com/terrastruct/d2/discussions/1581](url)

I'd love for any new port ability to support the following:
* Align port edges to a given top/bottom/left/right edge
* Specify distance from shape edge to port edge, including positive and negative values so you can have some overlap
* Distribute ports evenly along an edge with option to specify subset of edge length (ie not corner to corner; see below)
* Specify how connection leaves the port, eg orthogonal to edge or angle as needed
* Specify port shape
* Tool allows numbering/naming scheme for auto-naming of shapes including ports, eg CPU0, CPU1,...CPU255 or A0, A1, B0, B1, etc.

For alignment, I can envision a constraint graph that allows the following.  This wishlist is inspired by familiarity with 2D constraint-based sketch tools like Solidworks/Creo/Fusion, and frustration with Visio and other business-oriented, rather than engineering-oriented, tools:
* Make any set of edges colinear
* Make any set of shapes the same width or height
* Some addition to the DSL which allows an edge to be partitioned, for example you can refer to:
  * all of an edge
  * the top half of an edge
  * the inner 2/3 of an edge
  * the first 1/4 of an edge
  * etc
* This would allow you to align the size of a shape to some proportion of another shape, with proper alignment, or to distribute ports along only part of an edge, or to distribute one set of ports along one part of an edge, and another set of ports along another part of the same edge.

Obviously this is a lot of work, but i wanted to put this out there :)
thanks

**Comment by alixander** [comment]
@sonicsmooth Thank you for the kind words and detailed request!

I'm curious if you can share a diagram, maybe made by hand or another tool, of something you'd like the sum of these requests to achieve. The reason I ask this, is because I want to distill these features down to the least amount of higher order bits as possible. Often times, people ask for the capability to customize to 2^8 combination of values or something to achieve a much narrower combination of values they want. The overarching point of a tool like D2 is to simplify life and describe the system instead of describe the diagram. So maybe there's a specific diagram type or a directive you can help aggregate these features into such that the input just includes that one line and it can achieve the result you have in mind.

**Reply by sonicsmooth** [reply]
Sure,  
For example aligning edges will let you do this:
![image](https://github.com/user-attachments/assets/34ec75c1-eae7-4583-88ce-2fe3739fbbbf)
  
For port alignment, something like this.  Lots of details to sort out, probably not suitable for this high level post.  Specifically, "margin" outside of outermost port, and distance between port edges, or port pitch, etc.
![image](https://github.com/user-attachments/assets/54e7b068-3bec-44c7-a146-ab5660f3d25c)

When you combine, you get powerful output
![image](https://github.com/user-attachments/assets/deb14f78-d8f9-496f-b714-b49e91611b36)

Another modification.  Curves can have options of smooth curves or right angle routing.
![image](https://github.com/user-attachments/assets/45d73390-3637-4066-9064-88f8b9915e2f)

I'm not familiar enough currently with the syntax of D2 to suggest how to do the above.
There may be conditions where the graph is overconstrained and there would be an error message to the user.

Also it would be nice to be able to rotate and mirror shapes.
